zipline/gens/tradesimulation.py

```diff
@@ -206,46 +207,88 @@ def _process_snapshot(self, dt, snapshot, instant_fill):
 
-        for event in snapshot:
+        # Containers for the snapshotted events, so that the events are
+        # processed in a predictable order, without relying on the sorted order
+        # of the individual sources.
+
+        # There is only one benchmark per snapshot.
+        benchmark = None
+        # trades are initialized as a list since there is often a trade per
+        # snapshot, so better to allocate the list than check for is None
+        # on each event.
+        trades = []
+        # splits list will be allocated when a split occurs in the snapshot.
+        splits = None
+        # dividends list will be allocated when a dividend occurs in the
+        # snapshot.
+        dividends = None
+        # customs list will be allocated when a custom event (e.g. fetcher)
+        # occurs in the snapshot.
+        customs = None
 
+        for event in snapshot:
             if event.type == DATASOURCE_TYPE.TRADE:
-                self.update_universe(event)
-                any_trade_occurred = True
-                if instant_fill:
-                    events_to_be_processed.append(event)
-                else:
-                    for txn, order in blotter_process_trade(event):
-                        if txn.type == DATASOURCE_TYPE.TRANSACTION:
-                            perf_process_transaction(txn)
-                        elif txn.type == DATASOURCE_TYPE.COMMISSION:
-                            perf_process_commission(txn)
-                        perf_process_order(order)
-                    perf_process_trade(event)
-
+                trades.append(event)
             elif event.type == DATASOURCE_TYPE.BENCHMARK:
-                benchmark_event_occurred = True
-                perf_process_benchmark(event)
-                for txn, order in blotter_process_benchmark(event):
+                benchmark = event
+            elif event.type == DATASOURCE_TYPE.SPLIT:
+                if splits is None:
+                    splits = []
+                splits.append(event)
+            elif event.type == DATASOURCE_TYPE.CUSTOM:
+                if customs is None:
+                    customs = []
+                customs.append(event)
+            elif event.type == DATASOURCE_TYPE.DIVIDEND:
+                if dividends is None:
+                    dividends = []
+                dividends.append(event)
+            else:
+                raise log.warn("Unrecognized event=%s".format(event))
+
+        # Handle benchmark first.
+        #
+        # Internal broker implementation depends on the benchmark being
+        # processed first so that transactions and commissions reported from
+        # the broker can be injected.
+        if benchmark is not None:
+            benchmark_event_occurred = True
+            perf_process_benchmark(benchmark)
+            for txn, order in blotter_process_benchmark(benchmark):
+                if txn.type == DATASOURCE_TYPE.TRANSACTION:
```

Comment:
I'm not 100% why it was done this. I believe the intent was a way to splice in getting information back from the broker blotter into the tradesimulation loop, @richafrank might be able to give more context.

With the course of changing the tradesimulation loop to be focused on the simulation dt as top loop, we may want to change how the blotter is queried.

This dependency on getting the injection of values from the benchmark is impetus behind this patch, aiming to provide a setup where that timing of asking of the broker-backed blotter can be independent stream ordering.